test/ServiceTest.cpp

```diff
@@ -0,0 +1,107 @@
+#include "absl/strings/str_split.h"
+#include "ctre/ctre.h"
+#include "engine/Service.h"
+#include "parser/GraphPatternOperation.h"
+#include "util/http/HttpUtils.h"
+
+// _____________________________________________________________________________
+TEST(ServiceTest, computeResult) {
+  // Query execution context (with small test index) and allocator for testing,
+  // see `IndexTestHelpers.h`.
+  using ad_utility::AllocatorWithLimit;
+  QueryExecutionContext* testQec = ad_utility::testing::getQec();
+  AllocatorWithLimit<Id> testAllocator = ad_utility::testing::makeAllocator();
+
+  // Create and run a HTTP server, which accepts queries of exactly the
+  // following form, as a POST request, with content-type header
+  // `application/sparql-query` and accept header `text/tab-separated-values`:
+  //
+  // SELECT <list of variables, separated by space> WHERE {
+  //   <result of query as TSV, without header row>
+  // }
+  //
+  // It then returns a TSV which consists of a header row (created from the
+  // SELECT variables), followed by the TSV from the query body. Understand why
+  // the test below contains a TSV result in the query body, instead of a proper
+  // SPARQL query body: We don't want to test whether our test server can
+  // evaluate a real SPARQL query, we want to test whether the SERVICE operation
+  // does the right thing with the TSV sent from a remote endpoint. Note that
+  // the `SELECT ... WHERE` part is added by the `Service` operation, hence we
+  // do want to test that part.
+  //
+  // If the query does not exactly have the form described above, the test
+  // fails.
+  TestHttpServer httpServer([](auto request,
+                               auto&& send) -> boost::asio::awaitable<void> {
+    EXPECT_EQ(request.method(), http::verb::post);
+    EXPECT_EQ(request.base()[http::field::content_type],
+              "application/sparql-query");
+    EXPECT_EQ(request.base()[http::field::accept], "text/tab-separated-values");
+
+    // Parse the query.
+    std::string_view testSparqlQuery = request.body();
+    static constexpr auto testQueryRegex = ctll::fixed_string(
+        "^\\s*SELECT\\s+(.*)\\s+WHERE\\s*\\{\\s*((.*)\\n)\\s*\\}$");
+    auto match = ctre::search<testQueryRegex>(testSparqlQuery);
+    EXPECT_TRUE(match);
+    // Why do I need `template` before `get<...>` in the following? I don't need
+    // it in other places in the code (e.g, in `util/GeoSparqlHelpers.cpp`).
+    std::string_view variables = match.template get<1>();
+    std::string_view body = match.template get<2>();
+    std::string response = absl::StrCat(
+        absl::StrJoin(absl::StrSplit(variables, " "), "\t"), "\n", body);
+    // std::cout << "Reponse (TSV): " << std::endl << response;
+    co_return co_await send(ad_utility::httpUtils::createOkResponse(
+        response, request, ad_utility::MediaType::tsv));
+  });
+  httpServer.runInOwnThread();
+  Iri serviceIri{absl::StrCat("<http://localhost:", httpServer.getPort(), ">")};
```

Comment:
You should definitely write this test without using the Server.
Given your implementation of `Service` (it takes a plain function pointer and is not templated) you would have to
write a free function here with the correct interface, and to specify the test query for multiple cases you would have to make it access a (file-scoped) local variable.

The alternative is to store not a plain function pointer in the Service class, but a `std::function<istringstream(whatEverTheArguments)>`. This can hold a plain function pointer as well as a capturing lambda, then you can also work with a lambda that takes the query string and returns a lambda here.

(The overhead of `std::function` is ok for the `Service` class, since we have to send HttpRequests afterwards.